jupyterlab_gcedetails/src/components/auth_token_retrieval.tsx

```diff
@@ -1,14 +1,20 @@
-export function AuthTokenRetrieval(callback) {
+export function AuthTokenRetrieval(callback: (err, credentials) => void) {
   // eslint-disable-next-line prefer-const
   let authPopup: any;
+  const authOrigin = 'https://jupyterlab-interns-sandbox.uc.r.appspot.com';
   const listener = (message: MessageEvent) => {
-    authPopup.close();
-    window.removeEventListener('message', listener);
-    callback(message.data['error'], message.data['credentials']);
+    if (
+      message.origin === authOrigin &&
+      (message.data['error'] || message.data['credentials'])
+    ) {
+      authPopup.close();
+      window.removeEventListener('message', listener);
+      callback(message.data['error'], message.data['credentials']);
```

Comment:
What do you think about writing this as a Promise returning function? It could return `Promise<string>`. When the message was received from the auth site, it would check if `data.error` was set, and if not, call the `.resolve()` method on the promise. Otherwise, it would call `.reject()` with the error that was returned. Then, you could use `async/await` in the component that uses this function.

That would be consistent with how most of the other APIs we interact with work.